courses/Ada_For_The_C_Embedded_Developer/chapters/02_Perspective.rst

```diff
@@ -0,0 +1,802 @@
+
+We can actually simplify the Ada for loop a little further using iterators:
+
+.. code:: ada
+
+   with Ada.Text_IO; use Ada.Text_IO;
+
+   procedure Main
+   is
+      type Int_Array is array (Natural range 1 .. 100) of Integer;
+
+      List : Int_Array;
+   begin
+
+      for I in reverse List'Range loop
+         List (I) := List'Last - I;
+      end loop;
+
+      for I of List loop
+         Put (I'Img & " ");
+      end loop;
+
+      New_Line;
+   end Main;
+
+In the second for loop, we changed the syntax to :ada:`for I of List`. Instead of I being the index counter, it is now an iterator that references the underlying element. This example of Ada code is identical to the last bit of Ada code. We just used a different method to index over the second for loop. There is no C equivalent to this Ada feature, but it is similar to C++'s range based for loop.
+
+Type System
+--------------
+
+Ada is considered a "strongly typed" language. This means that the language does not define any implicit type conversions. C does define implicit type conversions, sometimes referred to as
+*integer promotion*. The rules for promotion are fairly straightforward in simple expressions but can get confusing very quickly. Let's look at a typical place of confusion with implicit type conversion:
+
+.. code:: c
+
+   !main.c
+   #include <stdio.h>
+
+   int main()
+   {
+      unsigned char a = 0xFF;
+      char b = 0xFF;
+
+      printf("Does a == b?\n");
+      if(a == b)
+         printf("Yes.\n");
+      else
+         printf("No.\n");
+
+      printf("a: 0x%08X, b: 0x%08X\n", a, b);
+
+      return 0;
+   }
+
+Run the above code. You will notice that :c:`a != b`! If we look at the output of the last :c:`printf` statement we will see the problem. :c:`a` is an unsigned number where :c:`b` is a signed number. We stored a value of :c:`0xFF` in both variables, but :c:`a` treated this as the decimal number :c:`255` while b treated this as the decimal number :c:`-1`. When we compare the two variables, of course they aren't equal; but thats not very intuitive. Lets look at the equivalent Ada example:
+
+.. code:: ada
+   :class: ada-expect-compile-error
+
+   with Ada.Text_IO; use Ada.Text_IO;
+
+   procedure Main
+   is
+      type Char is range 0 .. 255;
+      type Unsigned_Char is mod 256;
+
+      A : Char := 16#FF#;
+      B : Unsigned_Char := 16#FF#;
+   begin
+
+      Put_Line ("Does A = B?");
+
+      if A = B then
+         Put_Line ("Yes");
+      else
+         Put_Line ("No");
+      end if;
+
+   end Main;
+
+If you try to run this Ada example you will get a compilation error. This is because the compiler is telling you that you cannot compare variables of two different types. We would need to explicitly cast one side to make the comparison against two variables of the same type. By enforcing the explicit cast we can't accidentally end up in a situation where we assume something will happen implicitly when, in fact, our assumption is incorrect.
```

Comment:
At this point, some C-developers argue that a good C developer always pays attention and doesn't make this kind of mistakes. We should therefore emphasize that, although it's easy to spot the error in simple examples like this one, it becomes more complicated to find those bugs in larger code bases.